### PR TITLE
New `score_transition_risk_and_polish()` wraps and generalizes the "bundesbank" script

### LINKED ISSUE #200: New internal-ish helper to remove `co2e*` columns then `pivot_wider()`
Relates to https://github.com/2DegreesInvesting/tiltIndicatorAfter/issues/181
Relates to https://github.com/2DegreesInvesting/tiltWorkflows/issues/139

I understand the script @kalashsinghal used is the one at https://github.com/2DegreesInvesting/tiltIndicatorAfter/issues/153#issuecomment-2004469926 (except with the latest files form [tiltOuput-v0.0.0.9002](https://github.com/2DegreesInvesting/tiltOutput/releases) -- i.e. 03-sector-profile.zip and 06-emissions-profile.zip).

To sync the deleting the `co2e*` columns (1.) and widening the data (2.), I would suggest doing both things via a helper like this:

```r
library(dplyr)

remove_co2e_columns_then_pivot_wider <- function(data, id_cols) {
  # Every wide dataset should also lack the `co2e*` columns so it's best to do both steps in a single helper
  data |>
    # 1. Delete
    select(-matches("co2e_")) |>
    distinct() |>
    # 2. Widen
    tidyr::pivot_wider(
      id_cols = id_cols,
      names_from = emission_profile,
      names_prefix = "emission_category_",
      values_from = emission_profile_share
    )
}
```

I believe a good place for this code would be a developer-oriented helper (maybe hidden via `# @keywords internal`). A more user-facing location (e.g. in a workflow in tiltWorkflows like [this one](https://github.com/2DegreesInvesting/tiltWorkflows/issues/139)) doesn't seem like a good fit, since the wider format is not general but rather a specific requirement of only one bank.

### PR DESCRIPTION
**Pull request by maurolepore**
Closes #200 

This PR adds the new function `score_transition_risk_and_polish()`. Naming this function is hard because it's hard to name what it does other than wrapping the existing `score_transition_risk()`. Basically it makes the output readier for delivery, but that's pretty much the entire goal of tiltIndicatorAfter.

Note this function is internal, mostly to avoid making it too visible until we discuss if this package is a good place for it. If we do decide it stays here, then the existing `score_transition_risk()` should likely become internal and instead this function should become public.

<details/>
<summary/>
reprex
</summary>

``` r
devtools::load_all()
#> ℹ Loading tiltIndicatorAfter

library(readr, warn.conflicts = FALSE)
library(dplyr, warn.conflicts = FALSE)

withr::local_seed(123)
withr::local_options(
  readr.show_col_types = FALSE,
  tiltIndicatorAfter.output_co2_footprint = TRUE
)

toy_emissions_profile_products_ecoinvent <- read_csv(toy_emissions_profile_products_ecoinvent())
toy_emissions_profile_any_companies <- read_csv(toy_emissions_profile_any_companies())
toy_sector_profile_any_scenarios <- read_csv(toy_sector_profile_any_scenarios())
toy_sector_profile_companies <- read_csv(toy_sector_profile_companies())
toy_europages_companies <- read_csv(toy_europages_companies())
toy_ecoinvent_activities <- read_csv(toy_ecoinvent_activities())
toy_ecoinvent_europages <- read_csv(toy_ecoinvent_europages())
toy_ecoinvent_inputs <- read_csv(toy_ecoinvent_inputs())
toy_isic_name <- read_csv(toy_isic_name())

emissions_profile <- profile_emissions(
  companies = toy_emissions_profile_any_companies,
  co2 = toy_emissions_profile_products_ecoinvent,
  europages_companies = toy_europages_companies,
  ecoinvent_activities = toy_ecoinvent_activities,
  ecoinvent_europages = toy_ecoinvent_europages,
  isic = toy_isic_name
)
#> ℹ Adding 37% and 97% noise to `co2e_lower` and `co2e_upper`, respectively.

sector_profile <- profile_sector(
  companies = toy_sector_profile_companies,
  scenarios = toy_sector_profile_any_scenarios,
  europages_companies = toy_europages_companies,
  ecoinvent_activities = toy_ecoinvent_activities,
  ecoinvent_europages = toy_ecoinvent_europages,
  isic = toy_isic_name
)

result <- score_transition_risk_and_polish(emissions_profile, sector_profile)

result |> unnest_product()
#> # A tibble: 1,824 × 22
#>    companies_id          country main_activity ep_product activity_uuid_produc…¹
#>    <chr>                 <chr>   <chr>         <chr>      <chr>                 
#>  1 asteria_megalotomusq… austria wholesaler    tent       76269c17-78d6-420b-99…
#>  2 asteria_megalotomusq… austria wholesaler    tent       76269c17-78d6-420b-99…
#>  3 asteria_megalotomusq… austria wholesaler    tent       76269c17-78d6-420b-99…
#>  4 asteria_megalotomusq… austria wholesaler    tent       76269c17-78d6-420b-99…
#>  5 asteria_megalotomusq… austria wholesaler    tent       76269c17-78d6-420b-99…
#>  6 asteria_megalotomusq… austria wholesaler    tent       76269c17-78d6-420b-99…
#>  7 asteria_megalotomusq… austria wholesaler    tent       76269c17-78d6-420b-99…
#>  8 asteria_megalotomusq… austria wholesaler    tent       76269c17-78d6-420b-99…
#>  9 asteria_megalotomusq… austria wholesaler    tent       76269c17-78d6-420b-99…
#> 10 asteria_megalotomusq… austria wholesaler    tent       76269c17-78d6-420b-99…
#> # ℹ 1,814 more rows
#> # ℹ abbreviated name: ¹​activity_uuid_product_uuid
#> # ℹ 17 more variables: matched_activity_name <chr>,
#> #   matched_reference_product <chr>, unit <chr>, co2e_lower <dbl>,
#> #   co2e_upper <dbl>, emission_profile <chr>, benchmark <chr>,
#> #   profile_ranking <dbl>, tilt_sector <chr>, tilt_subsector <chr>,
#> #   co2_footprint <dbl>, sector_profile <chr>, scenario <chr>, year <dbl>, …

# Most banks need company-level results in long format
long <- result |> unnest_company() |> relocate(matches("emission_"))
long
#> # A tibble: 6,912 × 15
#>    emission_profile emission_profile_share companies_id    country main_activity
#>    <chr>                             <dbl> <chr>           <chr>   <chr>        
#>  1 high                                  1 asteria_megalo… austria wholesaler   
#>  2 high                                  1 asteria_megalo… austria wholesaler   
#>  3 high                                  1 asteria_megalo… austria wholesaler   
#>  4 high                                  1 asteria_megalo… austria wholesaler   
#>  5 medium                                0 asteria_megalo… austria wholesaler   
#>  6 medium                                0 asteria_megalo… austria wholesaler   
#>  7 medium                                0 asteria_megalo… austria wholesaler   
#>  8 medium                                0 asteria_megalo… austria wholesaler   
#>  9 low                                   0 asteria_megalo… austria wholesaler   
#> 10 low                                   0 asteria_megalo… austria wholesaler   
#> # ℹ 6,902 more rows
#> # ℹ 10 more variables: co2e_lower <dbl>, co2e_upper <dbl>, benchmark <chr>,
#> #   profile_ranking_avg <dbl>, co2_avg <dbl>, scenario <chr>, year <chr>,
#> #   reduction_targets_avg <dbl>, benchmark_tr_score_avg <chr>,
#> #   transition_risk_score_avg <dbl>

# Some banks need company-level results in wide format
wide <- score_transition_risk_and_polish(
  emissions_profile,
  sector_profile,
  pivot_wider = TRUE
) |>
  unnest_company() |> 
  relocate(matches("emission_"))
wide
#> # A tibble: 1,728 × 15
#>    emission_category_high emission_category_medium emission_category_low
#>                     <dbl>                    <dbl>                 <dbl>
#>  1                      1                        0                     0
#>  2                      1                        0                     0
#>  3                      1                        0                     0
#>  4                      1                        0                     0
#>  5                      1                        0                     0
#>  6                      1                        0                     0
#>  7                      1                        0                     0
#>  8                      1                        0                     0
#>  9                      1                        0                     0
#> 10                      1                        0                     0
#> # ℹ 1,718 more rows
#> # ℹ 12 more variables: emission_category_NA <dbl>, companies_id <chr>,
#> #   country <chr>, main_activity <chr>, benchmark <chr>,
#> #   profile_ranking_avg <dbl>, co2_avg <dbl>, scenario <chr>, year <chr>,
#> #   reduction_targets_avg <dbl>, benchmark_tr_score_avg <chr>,
#> #   transition_risk_score_avg <dbl>
```

</details>

----

TODO

- [x] [Link related issue/PR]([url](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)). 
- [x] Describe the goal of the PR. Avoid details that are clear in the diff.
- [x] Mark the PR as draft.
- [x] [Include a unit test](https://code-review.tidyverse.org/reviewer/aspects.html#sec-tests).
- [x] Review your own PR in "Files changed".
- [x] Ensure the PR branch is updated.
- [x] Ensure the checks pass.
- [x] Change the status from draft to ready.
- [x] Polish the PR title and description.
- [x] Assign a reviewer.

EXCEPTIONS

- [ ] Slide here any item that you intentionally choose to not do.

### CODE REVIEW

File: R/score_transition_risk_and_polish.R [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+          main_activity,
+          benchmark,
+          profile_ranking_avg,
+          co2_avg
+        ),
+        names_from = emission_profile,
+        names_prefix = "emission_category_",
+        values_from = emission_profile_share
+      )
+  }
```
Comment (kalashsinghal) [inline R/score_transition_risk_and_polish.R:183]: @maurolepore After applying `pivot_wider()` you didn't use `unchop(everything()) |>` command as I used in the script. I am not sure whether you missed it or excluded in by purpose. 

Nevertheless, I used it to convert all the list type columns into normal string type columns after `pivot_wider` create these list columns. All the four "emissions_category_" columns will become list columns by default, and I couldn't find a better solution to make them normal columns other than using the `unchop` function. So, please investigate and remove this issue maybe with a better solution. Thanks!

File: R/score_transition_risk_and_polish.R [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+      relationship = "many-to-many",
+      by = c("companies_id", "ep_product")
+    ) |>
+    mutate(benchmark_tr_score = paste(scenario, year, benchmark, sep = "_")) |>
+    left_join(select_transition_risk_score_at_product_level, by = c("companies_id", "ep_product", "benchmark_tr_score")) |>
+    distinct()
+    ## TODO: `europages_companies` should include headcount? Submit issue to tiltIndicatorBefore
+    # left_join(headcount, by = c("companies_id")) |>
+    ## This is specific to the latest bundesbank delivery
+    # filter(country == "germany", benchmark %in% c("unit", "unit_tilt_sector") | is.na(benchmark))
```
[the remark this answers is not in the document]
Comment (kalashsinghal) [reply R/score_transition_risk_and_polish.R:134]: @maurolepore Yes you can delete this code as it's specific to Anne's Bundesbank requirement.

File: R/score_transition_risk_and_polish.R [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    left_join(
+      select_sector_profile_at_product_level,
+      relationship = "many-to-many",
+      by = c("companies_id", "ep_product")
+    ) |>
+    mutate(benchmark_tr_score = paste(scenario, year, benchmark, sep = "_")) |>
+    left_join(select_transition_risk_score_at_product_level, by = c("companies_id", "ep_product", "benchmark_tr_score")) |>
+    distinct()
+    ## TODO: `europages_companies` should include headcount? Submit issue to tiltIndicatorBefore
+    # left_join(headcount, by = c("companies_id")) |>
```
[the remark this answers is not in the document]
Comment (kalashsinghal) [reply R/score_transition_risk_and_polish.R:132]: @maurolepore I agree that headcount is the `europages_companies` specification. And, yes we can add headcounts to this dataset, but then we also have to check whether the headcount columns create new errors in tiltIndicatorAfter while polishing product and company level results of four indicators. I would recommend you also create an issue in tiltIndicatorAfter for handling these columns well after we add them through tiltIndicatorBefore. Does it makes sense?

File: R/score_transition_risk_and_polish.R (hunk 1 above)
Comment (maurolepore) [reply R/score_transition_risk_and_polish.R:183]: FYI https://github.com/2DegreesInvesting/tiltIndicatorAfter/pull/206.

File: R/score_transition_risk_and_polish.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (maurolepore) [reply R/score_transition_risk_and_polish.R]: Good point. 

Note to self:

Here is an easy way to know. Create a local `toy_europages_companies()`  that overwrites the one from tiltToyData. If local `europages_companies`  has `min_headcount` and `max_eadcount` then we will know if the new columns cause any problem.

[the remark this answers is not in the document]
Comment (maurolepore) [reply R/score_transition_risk_and_polish.R]: FYI https://github.com/2DegreesInvesting/tiltIndicatorBefore/issues/97

File: tests/testthat/test-score_transition_risk_and_polish.R [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    ncol()
+  wide_cols <- wide |>
+    unnest_company() |>
+    select(matches("emission")) |>
+    ncol()
+  expect_true(long_cols < wide_cols)
+})
+
+test_that("with `*.output_co2_footprint` unset, `pivot_wider = FALSE` yiels no error", {
+  unset <- NULL
```
Comment (kalashsinghal) [inline tests/testthat/test-score_transition_risk_and_polish.R:91]: Typo: yiels -> yields

File: tests/testthat/test-score_transition_risk_and_polish.R [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    ecoinvent_activities = toy_ecoinvent_activities,
+    ecoinvent_europages = toy_ecoinvent_europages,
+    isic = toy_isic_name
+  )
+
+  expect_error(
+    score_transition_risk_and_polish(emissions_profile, sector_profile),
+    "tiltIndicatorAfter.output_co2_footprint"
+  )
+})
```
Comment (kalashsinghal) [inline tests/testthat/test-score_transition_risk_and_polish.R:126]: The title of the test says `yields no error`, however here we are expecting error. Am I missing something?

File: tests/testthat/test-score_transition_risk_and_polish.R [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+test_that("with `*.output_co2_footprint` unset, `pivot_wider = TRUE` yiels an error", {
+  unset <- NULL
```
Comment (kalashsinghal) [inline tests/testthat/test-score_transition_risk_and_polish.R:129]: Typo: yiels -> yields

File: tests/testthat/test-score_transition_risk_and_polish.R [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+  )
+
+  expect_no_error(
+    score_transition_risk_and_polish(
+      emissions_profile,
+      sector_profile,
+      pivot_wider = TRUE
+    )
+  )
+})
```
Comment (kalashsinghal) [inline tests/testthat/test-score_transition_risk_and_polish.R:167]: Here also, the description of the test says `yields an error`, however we are expecting no error inside the test. Please check

File: R/score_transition_risk_and_polish.R [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+          "main_activity",
+          "benchmark",
+          "profile_ranking_avg",
+          if (!pivot_wider) "co2_avg"
+        ),
+        names_from = "emission_profile",
+        names_prefix = "emission_category_",
+        values_from = "emission_profile_share"
+      )
+  }
```
Comment (kalashsinghal) [inline R/score_transition_risk_and_polish.R:188]: @maurolepore I think you also have to include the new argument `avoid_list_cols = TRUE` if we pivot wider. Isn't it?

Comment (kalashsinghal) [reply R/score_transition_risk_and_polish.R:188]: FYI: List columns are issue when we convert R dataframes into csv files.

File: tests/testthat/test-score_transition_risk_and_polish.R [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+    sector_profile,
+    pivot_wider = TRUE
+  ) |>
+    unnest_company() |>
+    select(matches("emission")) |>
+    lapply(typeof) |>
+    unlist() |>
+    unique()
+  expect_equal(type, "double")
+})
```
Comment (maurolepore) [inline tests/testthat/test-score_transition_risk_and_polish.R:211]: @kalashsinghal FYI, here's a new test to ensure the `emission*` columns are not list-columns.

**Pull request merged**